include/swift/AST/Attr.h

```diff
@@ -1024,34 +1027,84 @@ class SynthesizedProtocolAttr : public DeclAttribute {
   }
 };
 
+class AbstractFunctionDecl;
+
 /// The @_specialize attribute, which forces specialization on the specified
 /// type list.
 class SpecializeAttr : public DeclAttribute {
-  unsigned numTypes;
-  ConcreteDeclRef specializedDecl;
+public:
+  enum class SpecializationKind {
+    Full,
+    Partial
+  };
 
-  TypeLoc *getTypeLocData() {
-    return reinterpret_cast<TypeLoc *>(this + 1);
+private:
+  AbstractFunctionDecl *F;
```

Comment:
Yes. In general I didn't want to store it in the attribute. But in some contexts (e.g. in the printer) we currently don't pass the FuncDecl around. I'll see if it can done without too many changes.